### PR TITLE
Cache partition-type resolvability and cover dropped-source-column skip

### LINKED ISSUE #2844: Filtered scans fail entirely for tables where a historical partition spec's source column has been dropped
### Apache Iceberg Rust version

0.6.0 (latest version)

### Describe the bug

plan_files() returns an error for any scan with a predicate, even one that touches no partition column  if the snapshot contains a manifest written under a historical partition spec whose source column is no longer in the current schema. The unfiltered scan of the same table succeeds, so the data is readable; adding any filter makes the table's historical data unqueryable.

This is a legal table state in format v2 and later (including v3): a column may be dropped once it is no longer used by the default spec, but historical specs referencing it are retained in table metadata because existing manifests still use them.

Related: #2842 (the same table state panics the manifests metadata table; fix in #2843). This issue is the scan-planning counterpart and is independent of that fix.

### To Reproduce

1. Create a v2 table (v3 behaves identically) with schema `(id int, is_test boolean)` and spec `identity(is_test)` (spec 0), and write a snapshot so a manifest with spec 0 exists.
2. Evolve the spec to remove the `is_test partition` field (spec 1 becomes default).
3. Drop the `is_test column` from the schema.
4. Scan with any predicate on a live column, e.g. id > 0, and call plan_files().

Step 4 returns:

Unexpected => No column with source column id 2 in schema [...]

An unfiltered scan of the same table succeeds and returns the expected `FileScanTask`s (with the correct partition values, since manifest reading resolves types from the manifest's own embedded metadata).

### Expected behavior

The scan should succeed. Partition pruning is an optimization: a manifest whose spec can't be fully resolved should at worst not be pruned, never cause planning to fail. Two levels of fix:

1. Stopgap: when a spec's partition type can't be resolved against the current schema, skip partition pruning for that spec's manifests (treat the partition filter as always-true). This should be strictly safe since a predicate cannot reference a dropped column, so the skipped filter could never have pruned based on the missing field anyway; remaining row-level filtering is unaffected. I'd implement this as a resolvability pre-check (all of the spec's source IDs present in the schema) rather than by matching the error, so it composes with the longer-term fix.
2. Longer term: make partition_type succeed for this state: derive the output type from the transform where it doesn't depend on the source (bucket, year/month/day/hour, void) and degrade to the v3 unknown type only for identity/truncate. This is the approach iceberg-java is taking in apache/iceberg#17262 (context: apache/iceberg#16958); for rust it also depends on adding the v3 unknown type, which I'll file separately. With that in place, pruning keeps working on the still-resolvable fields and the stopgap becomes unnecessary.

I can contribute the stopgap fix.

### Willingness to contribute

I can contribute a fix for this bug independently

### PR DESCRIPTION
**Pull request by anoopj**
## Which issue does this PR close?

Closes #2844

## What changes are included in this PR?

Follow-up to #2845. The stopgap there computed partition_type once per manifest file in get_partition_filter to decide resolvability, duplicating the computation the partition filter cache already does on a miss. Move the fallback into the cache's miss path so an unresolvable spec yields a cached always-true filter keyed by spec id, and drop the per-manifest pre-check.

Also add a unit test for the constants_map skip on a dropped identity partition source column.

## Are these changes tested?

Yes, added a unit test

### CODE REVIEW

File: crates/iceberg/src/scan/cache.rs [hunk 1]
```diff
@@ -69,21 +69,33 @@ impl PartitionFilterCache {
...
+                );
+
+                let mut inclusive_projection = InclusiveProjection::new(partition_spec.clone());
+
+                inclusive_projection
+                    .project(&filter)?
+                    .rewrite_not()
+                    .bind(partition_schema.clone(), case_sensitive)?
+            }
+            Err(_) => BoundPredicate::AlwaysTrue,
```
Comment (laskoviymishka) [inline crates/iceberg/src/scan/cache.rs]: the `Err(_) => AlwaysTrue` here catches every failure `partition_type()` can return, not just the dropped-column case the comment is about. A schema-corruption bug or a future error path would get silently rewritten to always-true — and since we cache it by spec id, it's permanent for the rest of the scan, so there's no retry and nothing in the logs to explain why pruning quietly stopped.

I'd at minimum log a `tracing::warn!` with the `spec_id` and the error before falling back. Better still, narrow the arm to the known case — either a distinct `ErrorKind`, or an explicit `field_by_id(source_id).is_none()` check up front — and let genuinely unexpected errors keep propagating. wdyt?

File: crates/iceberg/src/arrow/record_batch_transformer.rs [hunk 2]
```diff
@@ -1484,6 +1484,83 @@ mod test {
...
     }
 
+    /// Test that a dropped identity partition source column is skipped rather than erroring.
+    ///
+    /// A historical spec may use `identity(col)` where `col` was later dropped from the
+    /// schema. `constants_map()` must skip that field instead of failing, so the file can
+    /// still be read (the dropped column is not projected).
+    #[test]
+    fn dropped_identity_partition_source_column_is_skipped() {
+        use crate::spec::{Struct, Transform};
```
Comment (laskoviymishka) [inline crates/iceberg/src/arrow/record_batch_transformer.rs:1494]: good regression guard — and since #2695 threads the real spec through the scan flow, the `constants_map` path this covers is live, not hypothetical. The gap is just that the change this PR actually makes — the `AlwaysTrue` fallback in `PartitionFilterCache::get` (cache.rs:97) — has no direct coverage. Nothing here builds a cache, calls `get()` with a spec whose source column was dropped, and asserts we get `AlwaysTrue` back and that it's cached.

Could we add a test that hits the cache path directly, or an integration test that scans a table with a predicate over a dropped partition column and asserts `plan_files()` succeeds? That's what we'd regress if someone reverted the `match`.

File: crates/iceberg/src/scan/cache.rs [hunk 3]
```diff
@@ -69,21 +69,33 @@ impl PartitionFilterCache {
...
-            .bind(partition_schema.clone(), case_sensitive)?;
+        // The partition type may fail to resolve against the schema. The known case is a
+        // historical spec whose source column was dropped, which is a legitimate v2+ state.
+        // Any resolution failure falls back to an always-true filter: files under the spec
+        // are not partition-pruned but still receive the row filter. The fallback is cached
+        // by spec id like any other filter.
+        // TODO(https://github.com/apache/iceberg-rust/issues/2844): derive partition types from
+        // transforms where possible to restore pruning on a historical spec's still-live fields,
+        // and narrow this to the dropped-column case once a distinguishable error exists.
+        let partition_filter = match partition_spec.partition_type(schema) {
```
Comment (laskoviymishka) [inline crates/iceberg/src/scan/cache.rs]: not blocking, and I see the TODO already flags this — just want to name the parity gap explicitly. Falling back to `AlwaysTrue` for the whole spec is coarser than Java/PyIceberg: they substitute `UnknownType` when a partition source field is missing, so `partition_type()` still resolves and `InclusiveProjection` returns always-true for only the dropped field's terms while live fields keep pruning. A spec like `bucket(N, dropped)`, `year(active_date)` still prunes on `active_date` in Java but loses all pruning here.

The real fix lives one level down in `partition_type()` (substitute the transform's fixed result type instead of erroring, which also covers the void-transform tombstone case), and that makes the fallback per-term for free. Fine to keep the spec-level stopgap for this PR — just worth making sure #2844 captures the per-term parity as the target.

File: crates/iceberg/src/scan/cache.rs [hunk 4]
```diff
@@ -69,21 +69,33 @@ impl PartitionFilterCache {
...
-
-        let partition_filter = inclusive_projection
-            .project(&filter)?
-            .rewrite_not()
-            .bind(partition_schema.clone(), case_sensitive)?;
+        // The partition type may fail to resolve against the schema. The known case is a
+        // historical spec whose source column was dropped, which is a legitimate v2+ state.
+        // Any resolution failure falls back to an always-true filter: files under the spec
+        // are not partition-pruned but still receive the row filter. The fallback is cached
+        // by spec id like any other filter.
```
Comment (laskoviymishka) [inline crates/iceberg/src/scan/cache.rs]: one thing worth a line of comment here: caching `AlwaysTrue` keyed by `spec_id` alone is only safe because this cache lives per-scan in `PlanContext` with a fixed schema and predicate. If it ever got hoisted to table or catalog scope, a spec cached as always-true couldn't be re-resolved by a later scan that would succeed. Could we note that per-scan assumption right where we cache, so a future refactor doesn't trip on it?

File: crates/iceberg/src/scan/context.rs [hunk 5]
```diff
@@ -175,23 +175,9 @@ impl PlanContext {
             .await
     }
 
-    /// Returns the partition filter for a manifest, or an always-true filter when the
-    /// manifest's spec cannot be resolved against the snapshot schema. Files of such
-    /// manifests are not partition-pruned but still receive the row filter.
     fn get_partition_filter(&self, manifest_file: &ManifestFile) -> Result<Arc<BoundPredicate>> {
```
Comment (laskoviymishka) [inline crates/iceberg/src/scan/context.rs:180]: small thing — the doc comment describing the always-true fallback came off with the pre-check, so `get_partition_filter` is now undocumented and the contract only lives in cache.rs. Could we leave a one-liner here pointing at `PartitionFilterCache::get` for the fallback behavior?

**Review by laskoviymishka — CHANGES_REQUESTED** [review]
Folding the fallback into `PartitionFilterCache::get` and removing the per-manifest `partition_type()` pre-check from `get_partition_filter` makes sense. It also removes the need for the #2845 workaround. I checked that the `constants_map` skip guard from #2845 is already on the branch, so the new transformer test passes.

I’d still hold the merge over the error handling.

`Err(_) => AlwaysTrue` catches every error from `partition_type()`, not only the dropped-column case. That means schema corruption or any future error path would be silently converted to `AlwaysTrue`, then cached for the rest of the scan. There would be no signal that anything went wrong.

At minimum, this should log a warning before falling back. Better yet, match only the known dropped-column error and let everything else propagate.

The new test also does not cover the cache fallback introduced here. It exercises the read-time `constants_map` path, so I’d add a direct test for the `PartitionFilterCache::get` fallback.

I left two smaller comments inline: one about documenting that the cache is per scan, and one about the Java per-term parity gap, which is fine to leave for #2844.

Once the fallback is narrowed and the cache path has direct coverage, I’ll take another look.

File: crates/iceberg/src/arrow/record_batch_transformer.rs (hunk 2 above)
Comment (anoopj) [reply crates/iceberg/src/arrow/record_batch_transformer.rs:1494]: Added dropped_partition_source_column_falls_back_to_always_true in cache.rs. It builds a cache and calls get() with a spec whose source column was dropped and assert we get `AlwaysTrue` and make sure if it's cached

File: crates/iceberg/src/scan/cache.rs (hunk 1 above)
Comment (anoopj) [reply crates/iceberg/src/scan/cache.rs]: I think that was a good call. I have narrowed it. Also added a warn with spec_id. I initially thought about a warn, but refrained because in my experience warn is almost always a copout. ;)

File: crates/iceberg/src/scan/cache.rs (hunk 3 above)
Comment (anoopj) [reply crates/iceberg/src/scan/cache.rs]: sounds good. Keeping spec-level stopgap for this PR. Updated the TODO to name the per-term parity

File: crates/iceberg/src/scan/cache.rs (hunk 4 above)
Comment (anoopj) [reply crates/iceberg/src/scan/cache.rs]: that is a good point. I incorporated most of your comment about into the comment itself.

File: crates/iceberg/src/scan/context.rs (hunk 5 above)
Comment (anoopj) [reply crates/iceberg/src/scan/context.rs:180]: Done.

File: crates/iceberg/src/scan/cache.rs [hunk 6]
```diff
@@ -69,21 +69,46 @@ impl PartitionFilterCache {
...
+        // partition type, so it falls back to an always-true filter: files under the spec
+        // are not partition-pruned but still receive the row filter. Any other resolution
+        // failure is unexpected and propagates. The fallback is cached by spec id like any
+        // other filter; this is safe only because the cache lives per-scan in `PlanContext`
+        // with a fixed schema and predicate. Hoisting it to table or catalog scope would
+        // pin a spec to always-true even for a later scan whose schema could resolve it.
+        // TODO(https://github.com/apache/iceberg-rust/issues/2844): derive partition types from
+        // transforms where possible to restore pruning on a historical spec's still-live fields,
+        // which also makes the fallback per-term (dropped fields only) instead of per-spec.
+        let dropped_source_column = partition_spec
```
Comment (laskoviymishka) [inline crates/iceberg/src/scan/cache.rs]: tiny thing — `dropped_source_column` reads as singular, but it's true when *any* field's source is dropped. `has_dropped_source_column` or `any_source_column_dropped` matches the `.any()` a bit better. Not blocking.

File: crates/iceberg/src/scan/cache.rs [hunk 7]
```diff
@@ -69,21 +69,46 @@ impl PartitionFilterCache {
...
+            .fields()
+            .iter()
+            .any(|field| schema.field_by_id(field.source_id).is_none());
 
-        let mut inclusive_projection = InclusiveProjection::new(partition_spec.clone());
+        let partition_filter = if dropped_source_column {
+            tracing::warn!(
```
Comment (laskoviymishka) [inline crates/iceberg/src/scan/cache.rs]: This fires before the write lock, so under the cache's benign double-check race (two threads missing the same `spec_id` at once) we'd emit the warn twice and insert AlwaysTrue twice. The insert is idempotent so it's harmless, but the duplicate log line is a new observable side effect that wasn't there before.

I'd move the warn onto the actual insert path so it only fires when we populate the entry. wdyt?

File: crates/iceberg/src/scan/cache.rs [hunk 8]
```diff
@@ -235,3 +260,81 @@ impl ExpressionEvaluatorCache {
...
+    use crate::expr::{Bind, BoundPredicate, Reference};
+    use crate::spec::{
+        Datum, FormatVersion, NestedField, PrimitiveType, Schema, SortOrder, TableMetadataBuilder,
+        Transform, Type, UnboundPartitionSpec,
+    };
+
+    /// A historical spec whose source column was dropped from the current schema resolves to an
+    /// always-true filter, and the fallback is cached under the spec id.
+    #[test]
+    fn dropped_partition_source_column_falls_back_to_always_true() {
```
Comment (laskoviymishka) [inline crates/iceberg/src/scan/cache.rs:279]: This exercises `cache.get` directly, which is exactly right for the caching assertion — but nothing drives the full `plan_files` → `get_partition_filter` → `get` path that the removed context.rs pre-check used to live on.

So if the plumbing in `get_partition_filter` regressed — wrong schema threaded through to `get`, say — this test would still pass. Could we add one scan-level test that scans a table with a historical dropped-column spec plus a predicate and asserts it completes and returns all rows? That's the path that actually changed here. wdyt?

**Review by laskoviymishka — APPROVED** [review]
This is good to land to me. 

Everything I flagged last round landed, and the refactor reads clean.

Left couple nit's and minor's in-line, i'll keep this PR soak for day or two before merging.

File: crates/iceberg/src/scan/cache.rs (hunk 6 above)
Comment (anoopj) [reply crates/iceberg/src/scan/cache.rs]: renamed.

File: crates/iceberg/src/scan/cache.rs (hunk 7 above)
Comment (anoopj) [reply crates/iceberg/src/scan/cache.rs]: yes, that _could_ help reduce the log volume. fixed.

File: crates/iceberg/src/scan/cache.rs (hunk 8 above)
Comment (anoopj) [reply crates/iceberg/src/scan/cache.rs:279]: There is a test that already does this? `test_filtered_scan_with_dropped_partition_source_column` it runs a filtered scan against a table whose manifest reference a spec with dropped source column and asserts the results. Adding a second test would duplicate it.
